### PR TITLE
Add memory / CPU usage properties to pipeline nodes

### LINKED ISSUE #1188: Allow different stages of a pipeline to run on differently-sized containers
In the pipeline in the `covid-notebooks` project (https://github.com/CODAIT/covid-notebooks/blob/master/pipelines/us_data.pipeline), one of the notebooks (`fit_us_data.ipynb`) takes considerably longer than the rest because it performs 10,000 curve-fitting operations serially, one after the other. I would like to make this notebook run faster by performing the curve-fitting operations in parallel.

This change will require having the `fit_us_data.ipynb` notebook run on a larger container with more cores. I would like for the remaining notebooks in the pipeline to continue running on smaller containers. 

Currently there is no way to specify in the pipeline editor UI that different stages of the pipeline should execute on containers with different amounts of CPU and/or memory. Containers inherit whatever is the default container size for the KFP cluster.

### PR DESCRIPTION
**Pull request by marthacryan**
Fixes #1188. Adds three fields to the properties editor in the pipeline editor to specify CPU, memory, and GPU usage for each node. The fields are optional. Memory has two fields: one to indicate a number, and another to indicate the unit of measurement (such as Gi, P, etc.) through a dropdown. There is validation to prevent the user from adding any negative values or any values over 99 for CPU and GPU (99 was chosen somewhat arbitrarily). 

Here's how they look with the small properties editor:
image

And with the large properties editor:
image

This is the dropdown for memory:
image

This is the validation message:
image

There is a bug on the canvas side that I want to investigate - the validation considers an empty field to be 0, so it shows the validation error above when the input field is empty. 

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by lresende** [comment]
Could we simplify and always use GB (as in updating the label) and that allows us to remove the unit field? Also, these will most of the time be single digits except for memory that would be maybe 2 or 3 digits, so we could try to use less real estate for the fields?

**Comment by marthacryan** [comment]
> Could we simplify and always use GB (as in updating the label) and that allows us to remove the unit field? Also, these will most of the time be single digits except for memory that would be maybe 2 or 3 digits, so we could try to use less real estate for the fields?

@lresende I can definitely just use GB to remove the unit field. I tried moving them to just one line and in the small size it looks a little crowded if you include GB in the title:
image
I could spread them over 2 lines:
image
Or it also looks a little less crowded if you move the "GB" out of the title of the RAM:
image

**Comment by kevin-bates** [comment]
Could we place the "PUs" adjacent to each other?  I understand RAM and CPU may be on the same line, depending on the window's width, but in the case where they're all vertically or horizontally aligned, I think having CPU and GPU adjacent is better from a UX perspective.

**Comment by marthacryan** [comment]
Updated look:
image

Does anyone have feedback / suggestions for the descriptions?
image
image
image

**Comment by akchinSTC** [comment]
that looks good!

**Comment by lresende** [comment]
Just let's be careful with decimals around GPUs (and maybe others) as in the past GPUs factions could not be assigned to a pod.

**Comment by marthacryan** [comment]
@lresende @akchinSTC I checked with Alan before allowing all fields to be doubles, and I believe he checked to make sure it'd work

**Comment by lresende** [comment]
This [page](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/)  says:

'Each container can request one or more GPUs. It is not possible to request a fraction of a GPU.'

**Comment by marthacryan** [comment]
> This [page](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/) says:
> 
> 'Each container can request one or more GPUs. It is not possible to request a fraction of a GPU.'

Oh ok so I'll set the GPU field to integer (unless Alan disagrees?) but keep the other 2 fields as doubles?

**Comment by marthacryan** [comment]
Realized that it isn't possible to ensure that a number field is strictly an integer - there isn't a way around it at the moment but I can add a comment in the description and I added an [issue](https://github.com/elyra-ai/canvas/issues/451) in canvas.

File: packages/pipeline-editor/src/properties.json [hunk 1]
```diff
@@ -18,6 +18,21 @@
...
       "required": true
     },
+    {
+      "id": "cpu",
+      "type": "double",
+      "required": "false"
+    },
+    {
+      "id": "gpu",
+      "type": "double",
```
Comment (lresende) [inline packages/pipeline-editor/src/properties.json:28]: do we need to set this as integer?

**Comment by lresende** [comment]
This LGTM now. @akchinSTC how are we planning to integrate the backend changes? Will it be part of this PR?

File: packages/pipeline-editor/src/properties.json (hunk 1 above)
Comment (marthacryan) [reply packages/pipeline-editor/src/properties.json:28]: In the future we will, but right now changing it to integer won't change any behaviors in canvas

**Comment by akchinSTC** [comment]
@lresende - yes im going to be adding the backend changes using this as a base.

**Comment by akchinSTC** [comment]
Dependent kfp-notebook PR here : https://github.com/elyra-ai/kfp-notebook/pull/78

**Comment by lresende** [comment]
Thanks @akchinSTC, I looked in the backend code and I was wondering how are we falling back to the current behavior of not sending these parameters if not set/selected by the user.

**Comment by kevin-bates** [comment]
I think this might warrant a pipeline version increment.  The problematic case being a new pipeline file with these parameters set, getting used by an older version - which would then save the pipeline w/o these values (I think).

**Comment by akchinSTC** [comment]
Used a new pipeline file with cpu/mem resources defined, on Elyra 1.x. Upon saving the properties in the node, I've confirmed it will not overwrite/delete the resource values.

**Comment by lresende** [comment]
A few issues:
 - Looks like when you set a value to a field (e.g. gpu) and then reset (e.g. select + delete) the field values are being set as null instead of something like an empty string or no field at all.
 - Looks like you can use the little buttons to increase/decrease the values, but it's going below 0, which it should not. 


And things working well:
 - I tested the pipeline being edited by the pipeline editor with these changes and see the fields there, going back to an old version of the editor and editing the pipeline does not seem to cause any side effects or erase the fields from the pipeline file, which means we don't need a version change for the pipeline. 


I will continue testing the runtime integration tomorrow, as the KFP server I was using was down when I was testing it.

**Comment by marthacryan** [comment]
> * Looks like when you set a value to a field (e.g. gpu) and then reset (e.g. select + delete) the field values are being set as null instead of something like an empty string or no field at all.

I think it wouldn't be possible to set it to empty string because it's a number field. I can add logic to check if the field is null and delete it from the properties (when applying properties) if that helps the backend, but would it be different in terms of the python code to have a field set to null vs. the field existing? @akchinSTC 

> * Looks like you can use the little buttons to increase/decrease the values, but it's going below 0, which it should not.

Unfortunately I think this is a behavior on the canvas side (I just reached out to them to make sure), I don't think there's a way around that at the moment, but it does show a validation error immediately when you press the button and it becomes an invalid field. I added [an issue](https://github.com/elyra-ai/canvas/issues/455) and they said that the behavior we're looking for already exists in the component they're using, they just have to add it as an option, so it should be relatively quick.

File: elyra/templates/kfp_template.jinja2 [hunk 2]
```diff
@@ -22,6 +22,9 @@ def create_pipeline():
                                                       cos_dependencies_archive='{{ operation.cos_dependencies_archive }}',
                                                       pipeline_inputs={{ operation.pipeline_inputs }},
                                                       pipeline_outputs={{ operation.pipeline_outputs }},
+                                                      cpu_request='{{ operation.cpu_request }}',
+                                                      mem_request='{{ operation.mem_request }}G',
```
Comment (marthacryan) [inline elyra/templates/kfp_template.jinja2:26]: Is there supposed to be a G at the end of this line?

Comment (akchinSTC) [reply elyra/templates/kfp_template.jinja2:26]: its a little weird I admit, looking for hands if we want to change it. 
We could add another parameter to the template rendering function in the kfp processor to pass the value with `G` appended to it, in this case it was just quicker to add a G to the template.

**Comment by akchinSTC** [comment]
Tests are failing due to the kfp-notebook changes that this PR is dependent on, have not been merged and new package pushed to pypi. In order to test,  

> Dependent kfp-notebook PR here : [elyra-ai/kfp-notebook#78](https://github.com/elyra-ai/kfp-notebook/pull/78)

pull this PR and install locally.

**Review by lresende — APPROVED** [review]

**Pull request merged**